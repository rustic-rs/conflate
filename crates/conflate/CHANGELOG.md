# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.2.0](https://github.com/rustic-rs/conflate/releases/tag/conflate-v0.2.0) - 2024-09-26

### Other

- fix naming in readmes
- change crate names to conflate*
- change crate names to conflate*
<!---
SPDX-FileCopyrightText: 2020 Robin Krahl <robin.krahl@ireas.org>
SPDX-License-Identifier: CC0-1.0
-->

# v0.1.0 (2020-09-01)

Initial release providing the `Merge` trait and some merge strategies in the
`bool`, `num`, `ord` and `vec` modules.
