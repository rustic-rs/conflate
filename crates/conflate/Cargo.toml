# SPDX-FileCopyrightText: 2020 Robin Krahl <robin.krahl@ireas.org>
# SPDX-License-Identifier: CC0-1.0

[package]
name = "conflate"
version = "0.3.0"
authors = ["the rustic-rs team", "Robin Krahl <robin.krahl@ireas.org>"]
categories = ["rust-patterns"]
documentation = "https://docs.rs/conflate"
edition = "2021"
exclude = [".builds/*"]
homepage = "https://github.com/rustic-rs/conflate"
keywords = ["merge", "macros", "derive", "conflate"]
license = "Apache-2.0 OR MIT"
readme = "README.md"
repository = "https://github.com/rustic-rs/conflate"
rust-version = { workspace = true }
description = "Merge multiple values into one"

[features]
default = ["derive", "num", "std"]
derive = ["dep:conflate_derive"]
num = ["dep:num-traits"]
std = []

[dependencies]
conflate_derive = { workspace = true, optional = true }
num-traits = { version = "0.2", optional = true }

[dev-dependencies]
trybuild = "1.0"
