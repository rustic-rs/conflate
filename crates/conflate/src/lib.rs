// SPDX-FileCopyrightText: 2020 Robin Krahl <robin.krahl@ireas.org>
// SPDX-License-Identifier: Apache-2.0 or MIT

//! Provides [`Merge`][], a trait for objects that can be merged.
//!
//! # Usage
//!
//! ```
//! trait Merge {
//!     fn merge(&mut self, other: Self);
//! }
//! ```
//!
//! The [`Merge`][] trait can be used to merge two objects of the same type into one.  The intended
//! use case is merging configuration from different sources, for example environment variables,
//! multiple configuration files and command-line arguments, see the [`args.rs`][] example.
//!
//! This crate does not provide any `Merge` implementations, but `Merge` can be derived for
//! structs.  When deriving the `Merge` trait for a struct, you can provide custom merge strategies
//! for the fields that donâ€™t implement `Merge`.  A merge strategy is a function with the signature
//! `fn merge<T>(left: &mut T, right: T)` that merges `right` into `left`.  The submodules of this
//! crate provide strategies for the most common types, but you can also define your own
//! strategies.
//!
//! ## Features
//!
//! This crate has the following features:
//!
//! - `derive` (default):  Enables the derive macro for the `Merge` trait using the `conflate_derive`
//!   crate.
//! - `num` (default): Enables the merge strategies in the `num` module that require the
//!   `num_traits` crate.
//! - `std` (default): Enables the merge strategies in the `hashmap` and `vec` modules that require
//!    the standard library.  If this feature is not set, `conflate` is a `no_std` library.
//!
//! # Example
//!
//! ```
//! use conflate::Merge;
//!
//! #[derive(Merge)]
//! struct User {
//!     // Fields with the skip attribute are skipped by Merge
//!     #[merge(skip)]
//!     pub name: &'static str,
//!
//!     // The strategy attribute is used to customize the merge behavior
//!     #[merge(strategy = conflate::option::overwrite_none)]
//!     pub location: Option<&'static str>,
//!
//!     #[merge(strategy = conflate::vec::append)]
//!     pub groups: Vec<&'static str>,
//! }
//!
//! let defaults = User {
//!     name: "",
//!     location: Some("Internet"),
//!     groups: vec!["rust"],
//! };
//! let mut ferris = User {
//!     name: "Ferris",
//!     location: None,
//!     groups: vec!["mascot"],
//! };
//! ferris.merge(defaults);
//!
//! assert_eq!("Ferris", ferris.name);
//! assert_eq!(Some("Internet"), ferris.location);
//! assert_eq!(vec!["mascot", "rust"], ferris.groups);
//! ```
//!
//! [`Merge`]: trait.Merge.html
//! [`args.rs`]: https://git.sr.ht/~ireas/merge-rs/tree/master/examples/args.rs

#![cfg_attr(not(feature = "std"), no_std)]

#[cfg(feature = "derive")]
pub use conflate_derive::*;

pub mod bool;
#[cfg(feature = "std")]
pub mod hashmap;
pub mod iter;
#[cfg(feature = "num")]
pub mod num;
pub mod option;
pub mod ord;
#[cfg(feature = "std")]
pub mod vec;

/// A trait for objects that can be merged.
///
/// # Deriving
///
/// `Merge` can be derived for structs if the `derive` feature is enabled.  The generated
/// implementation calls the `merge` method for all fields, or the merge strategy function if set.
/// You can use these field attributes to configure the generated implementation:
/// - `skip`: Skip this field in the `merge` method.
/// - `strategy = f`: Call `f(self.field, other.field)` instead of calling the `merge` function for
///    this field.
///
/// You can also set a default strategy for all fields by setting the `strategy` attribute for the
/// struct.
///
/// # Examples
///
/// Deriving `Merge` for a struct:
///
/// ```
/// use conflate::Merge;
///
/// #[derive(Debug, PartialEq, Merge)]
/// struct S {
///     #[merge(strategy = conflate::option::overwrite_none)]
///     option: Option<usize>,
///
///     #[merge(skip)]
///     s: String,
///
///     #[merge(strategy = conflate::bool::overwrite_false)]
///     flag: bool,
/// }
///
/// let mut val = S {
///     option: None,
///     s: "some ignored value".to_owned(),
///     flag: false,
/// };
/// val.merge(S {
///     option: Some(42),
///     s: "some other ignored value".to_owned(),
///     flag: true,
/// });
/// assert_eq!(S {
///     option: Some(42),
///     s: "some ignored value".to_owned(),
///     flag: true,
/// }, val);
/// ```
///
/// Setting a default merge strategy:
///
/// ```
/// use conflate::Merge;
///
/// #[derive(Debug, PartialEq, Merge)]
/// #[merge(strategy = conflate::option::overwrite_none)]
/// struct S {
///     option1: Option<usize>,
///     option2: Option<usize>,
///     option3: Option<usize>,
/// }
///
/// let mut val = S {
///     option1: None,
///     option2: Some(1),
///     option3: None,
/// };
/// val.merge(S {
///     option1: Some(2),
///     option2: Some(2),
///     option3: None,
/// });
/// assert_eq!(S {
///     option1: Some(2),
///     option2: Some(1),
///     option3: None,
/// }, val);
/// ```
pub trait Merge {
    /// Merge another object into this object.
    fn merge(&mut self, other: Self);
}
