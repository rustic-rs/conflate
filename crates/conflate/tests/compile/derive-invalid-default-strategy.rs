// SPDX-FileCopyrightText: 2020 Robin Krahl <robin.krahl@ireas.org>
// SPDX-License-Identifier: Apache-2.0 or MIT

use conflate::Merge;

#[derive(Merge)]
#[merge(strategy = my_custom_merge_strategy)]
struct S {
    field1: u16,
}

fn my_custom_merge_strategy(left: &mut u8, right: u8) {
    *left += right
}

fn main() {}
