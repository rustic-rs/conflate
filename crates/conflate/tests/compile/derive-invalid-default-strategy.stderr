error[E0308]: arguments to this function are incorrect
  --> tests/compile/derive-invalid-default-strategy.rs:7:20
   |
7  |   #[merge(strategy = my_custom_merge_strategy)]
   |  ____________________-^^^^^^^^^^^^^^^^^^^^^^^
8  | | struct S {
9  | |     field1: u16,
   | |          -
   | |          |
   | |__________expected `&mut u8`, found `&mut u16`
   |            expected `u8`, found `u16`
   |
   = note: expected mutable reference `&mut u8`
              found mutable reference `&mut u16`
note: function defined here
  --> tests/compile/derive-invalid-default-strategy.rs:12:4
   |
12 | fn my_custom_merge_strategy(left: &mut u8, right: u8) {
   |    ^^^^^^^^^^^^^^^^^^^^^^^^ -------------  ---------
help: you can convert a `u16` to a `u8` and panic if the converted value doesn't fit
   |
9  |     field1.try_into().unwrap(): u16,
   |           ++++++++++++++++++++
